### PR TITLE
Bugfix: `spack find -x` in environments

### PR DESCRIPTION
**Pull request by scheibelp**
This addresses part [1] of https://github.com/spack/spack/issues/46345

Before this PR, all query parameters like date ranges, `-x`, etc. were ignored when an env was active. With this PR:

```
$ spack find -x
==> In environment findx
==> 3 root specs
 -  libpng +pic   -  py-numpy  [+] python

-- darwin-ventura-m1 / apple-clang@14.0.3 -----------------------
python-venv@1.0
==> 1 installed package
==> 8 concretized packages to be installed (show with `spack find -c`)


$ spack find -X
==> In environment findx
==> 3 root specs
 -  libpng +pic   -  py-numpy  [+] python

-- darwin-ventura-m1 / apple-clang@14.0.3 -----------------------
apple-libuuid@1353.100.2            expat@2.5.0           libffi@3.4.4      openssl@3.1.3       py-packaging@23.1     readline@8.2   zstd@1.5.5
berkeley-db@18.1.40                 gdbm@1.23             libiconv@1.17     perl@5.38.0         py-pip@23.1.2         sqlite@3.43.2
bzip2@1.0.8                         gettext@0.22.4        libxcrypt@4.4.35  pigz@2.7            py-setuptools@68.0.0  tar@1.34
ca-certificates-mozilla@2023-05-30  gmake@4.4.1           libxml2@2.10.3    pkgconf@1.9.5       py-wheel@0.41.2       xz@5.4.1
diffutils@3.9                       gnuconfig@2022-09-17  ncurses@6.4       py-flit-core@3.9.0  python@3.11.6         zlib-ng@2.1.5
==> 31 installed packages
==> 8 concretized packages to be installed (show with `spack find -c`)
```

Note: for concretized-but-not-installed specs, query parameters are generally not applicable, so they are not filtered.

Example of before:

```
$ spack find -x
==> In environment findx
==> 3 root specs
 -  libpng +pic   -  py-numpy  [+] python

-- darwin-ventura-m1 / apple-clang@14.0.3 -----------------------
apple-libuuid@1353.100.2            expat@2.5.0           libffi@3.4.4      openssl@3.1.3       py-packaging@23.1     python-venv@1.0  zlib-ng@2.1.5
berkeley-db@18.1.40                 gdbm@1.23             libiconv@1.17     perl@5.38.0         py-pip@23.1.2         readline@8.2     zstd@1.5.5
bzip2@1.0.8                         gettext@0.22.4        libxcrypt@4.4.35  pigz@2.7            py-setuptools@68.0.0  sqlite@3.43.2
ca-certificates-mozilla@2023-05-30  gmake@4.4.1           libxml2@2.10.3    pkgconf@1.9.5       py-wheel@0.41.2       tar@1.34
diffutils@3.9                       gnuconfig@2022-09-17  ncurses@6.4       py-flit-core@3.9.0  python@3.11.6         xz@5.4.1
==> 32 installed packages
==> 8 concretized packages to be installed (show with `spack find -c`)

$ spack find -X
(exact same output as previous)
```

### CODE REVIEW

**Comment by haampie** [comment]
There are still tons of problems.

* Positional arguments `spack find x` influence "concretized packages to be installed", but flags like `--explicit` and `--tags` do not (for different reasons).

* `--only-missing`, `--missing`, `--only-deprecated`, `--deprecated` do not work because you're branching on `.installed`. (keep the meaning of `--missing` specs in the database that have refcount > 0 but are not installed, please don't conflate with the `-c` flag)

* `spack -e . find -c uninstalled-root` errors.

* minor: `spack -e . find not-installed-or-in-env` errors but `spack -e find --start-date <in the future>` does not, which is inconsistent.

Looks like you're branching way too early on this `-c` option. 

Instead collect both `specs_known_to_db = spack.store.STORE.db.query(hashes=env_hashes, installed=any)` and `specs_matching_query = spack.store.STORE.db.query(hashes=env_hashes, **q_args)` and assemble `results` as the subset of env specs that are either in `specs_matching_query` or not in `specs_known_to_db`. Then make lack of `-c` the last filter on `results` where specs not known to the db are removed. That should ensure all filters are applied consistently, except you may want to have "missing specs" (i.e. uninstalled specs with refcount > 0 in the db) still pop up under the "concretized packages to be installed" bit, which should be a matter of getting `installed=*` right in `specs_known_to_db`.

**Comment by scheibelp** [comment]
Thanks for listing out those cases. This is now updated and tested.

> Positional arguments spack find x influence "concretized packages to be installed", 

I assume we agree they should.

> but flags like --explicit 

Record-based query params don't apply to concretized-but-not-installed specs, so the answer has to be either 0 or just whatever is in the env.

`--loaded` is not a record parameter, but IMO it doesn't make sense to apply as a filter to concretized-but-not-installed specs, because it effectively would remove all of them (unless someone was doing something strange with their env, and I don't think `spack find` is the tool to help them fix that).

> and --tags do not (for different reasons).

The `--tags` part is fixed now (unlike `--explicit`, it isn't record based, so IMO does make sense to filter env specs on)

> --only-missing, --missing, --only-deprecated, --deprecated do not work because you're branching on .installed. (keep the meaning of --missing specs in the database that have refcount > 0 but are not installed, please don't conflate with the -c flag)

I now test `--missing` (and the logic doesn't center around `.installed` for the default query results now.

> spack -e . find -c uninstalled-root errors.

That is fixed now (and tested)

> spack -e . find not-installed-or-in-env errors but spack -e find --start-date <in the future> does not, which is inconsistent.

I think that's how it has worked before. This sounds more like a desire to change behavior. I'd like to focus on things that are definitely bugs in this PR.

File: lib/spack/spack/cmd/find.py [hunk 1]
```diff
@@ -299,43 +297,66 @@ def root_decorator(spec, string):
...
         args.groups = not args.format
 
     # Exit early with an error code if no package matches the constraint
-    if not results and args.constraint:
+    if concretized_but_not_installed and args.show_concretized:
+        pass
+    elif results:
+        pass
+    elif args.constraint:
         constraint_str = " ".join(str(s) for s in args.constraint_specs)
```
Comment (haampie) [inline lib/spack/spack/cmd/find.py:333]: can you fix `constraint_specs` as well

File: lib/spack/spack/cmd/find.py [hunk 2]
```diff
@@ -299,43 +297,66 @@ def root_decorator(spec, string):
...
         tty.die(f"No package matches the query: {constraint_str}")
```
Comment (haampie) [inline lib/spack/spack/cmd/find.py]: Can you fix this as well then?

```
$ spack find x
==> Error: No package matches the query: x

$ spack -e . find  x
==> Error: No package matches the query:   # <-- what query
```

Comment (haampie) [reply lib/spack/spack/cmd/find.py]: Also better to make it raise than tty.die if you meant to isolate the function for unit tests.

File: lib/spack/spack/main.py [hunk 3]
```diff
@@ -692,6 +692,30 @@ def _log_command_output(self, out):
                     tty.verbose(fmt.format(ln.replace("==> ", "")))
 
 
+class SpackCommandArgs:
```
Comment (haampie) [inline lib/spack/spack/main.py]: Can you move it into lib/spack/spack/test please.

**Comment by tgamblin** [comment]
@haampie can you look at this one again?

File: lib/spack/spack/cmd/find.py [hunk 4]
```diff
@@ -299,43 +297,70 @@ def root_decorator(spec, string):
...
         else:
-            results = env.all_specs()
+            env_specs = all_env_specs
+
+        spec_hashes = set(x.dag_hash() for x in env_specs)
+        specs_meeting_q_args = set(spack.store.STORE.db.query(hashes=spec_hashes, **q_args))
+
+        results = list()
+        for spec in env_specs:
+            if not spec.installed:
```
Comment (haampie) [inline lib/spack/spack/cmd/find.py]: Should be wrapped in a read transaction.

**Review by haampie — APPROVED** [review]
Approving this cause it fixes bugs, but not convinced `spack find` is currently great.

**Pull request merged**